### PR TITLE
Optimise rge matrix multiplication

### PR DESCRIPTION
**Pull request by LucaMantani**
This PR aims at optimising the theory predictions when RGE matrices are included.
The main idea is to exploit the following equation

Screenshot 2024-11-21 at 19 36 31

redefining the linear correction matrix and the quadratic correction matrix by absorbing the RGE matrix Gamma into them. Since this operation needs to be done only once at the beginning, the theory prediction evaluation later on becomes much faster.

**NOTE**: while the linear correction matrix was already in that form, the quadratic was not produced in that way before. I changed the way it is defined now in order to conform to that expression, which I find also more intuitive and makes the code more streamlined. In particular, the quadratic correction matrix is defined as a triangular matrix. One could have decided instead to make it a symmetric matrix filling both Op1xOp2 and Op2xOp1 with half the number we have in the theory tables but I preferred the option of the triangular matrix.

The fits with RGE are now much faster, e.g. the fits in the following reports are identical:
[report_example.pdf](https://github.com/user-attachments/files/17850475/report_example.pdf)

the one in main took 32 minutes, while the one in this new branch took 4 minutes.

Another considerable advantage of this approach is that it's going to make the adaptation of other parts of the code to account for RGE easier. For example, in the analytic fit routine, one will not need to modify too much the code since the linear correction matrix will have the rge embedded automatically (if loaded). The same is true for the Fisher information matrix.

### CODE REVIEW

File: src/smefit/compute_theory.py
Comment (giacomomagni) [inline src/smefit/compute_theory.py:1]: Also the function `flatten` now can be dropped.

**Review by giacomomagni — COMMENTED** [review]
Thank you @LucaMantani, indeed this implementation is more clean!! 

I have two small suggestions: 

* maybe you can considering adding a check/test that the Quad corrections are always upper triangular after the loader. 
* did you check that the fits with constrained coefficients (in particular non linear ones) are still working?  I.e that all the cases where `QuadraticCorrections` was used are now in good shape ?

File: src/smefit/compute_theory.py
Comment (LucaMantani) [reply src/smefit/compute_theory.py:1]: True, I will remove it.

**Comment by LucaMantani** [comment]
> Thank you @LucaMantani, indeed this implementation is more clean!! 
> 
> 
> 
> I have two small suggestions: 
> 
> 
> 
> * maybe you can considering adding a check/test that the Quad corrections are always upper triangular after the loader. 
> 
> * did you check that the fits with constrained coefficients (in particular non linear ones) are still working?  I.e that all the cases where `QuadraticCorrections` was used are now in good shape ?  Or you were planning to do it in #93, which now seems superseded by this PR  ?
> 
> 
> 
> 

Thanks for these comments.

1) I will add a test checking that.

2) Do you have an example I could test to check this? The Fisher PR is indeed superseded and it will need to be readapted but it should be relatively easy with the new loader.

**Comment by LucaMantani** [comment]
if there are tests, they pass, but I will check that explicitly now.

Regarding the Fisher, it has now the quadratic in a good shape but it's not implementing the RGE. One will need to read the matrices when doing reports and then they will indeed be included automatically by the Loader.

**Comment by giacomomagni** [comment]
> Regarding the Fisher, it has now the quadratic in a good shape but it's not implementing the RGE. One will need to read the matrices when doing reports and then they will indeed be included automatically by the Loader.

I'm confused... the  report is already using the Loader, so all the RGE effect will be included with tis PR no ?
Or you mean other multiplication have to be done ?

File: src/smefit/loader.py [hunk 1]
```diff
@@ -509,36 +512,78 @@ def construct_corrections_matrix(corrections_list, n_data_tot, sorted_keys=None)
...
+
+    if rgemat is not None:
+        if len(rgemat.shape) == 3:
+            corr_values = jnp.einsum(
+                "ijk, ijl, ikr -> ilr", corr_values, rgemat, rgemat
+            )
+        else:
+            corr_values = jnp.einsum("ijk, jl, kr -> ilr", corr_values, rgemat, rgemat)
+
+    return corr_values
```
Comment (jacoterh) [inline src/smefit/loader.py:588]: Without RG, this is upper triangular indeed. But this property is lost when the RG is included due to the jnp.einsum contraction above. Are we double counting now? Probably not, but I don't see why atm.

**Review by jacoterh — COMMENTED** [review]
Hi @LucaMantani , great we get such a nice speed up, it makes much more sense to do it this way. I added some extra comments in the code just to make sure we understand what we're doing in a year from now. I have one question regarding the triangular property after accounting the RG, see inline.

File: src/smefit/loader.py (hunk 1 above)
Comment (LucaMantani) [reply src/smefit/loader.py:588]: Mmm, good spot...

**Comment by LucaMantani** [comment]
I tested for a fake UV model with non-linear constraints (card used in attachment) and it works fine:
main:
```
┏━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━┓
┃ Parameter ┃ Best value ┃ Error ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━┩
│ g         │ 0.000      │ 0.332 │
└───────────┴────────────┴───────┘
```

branch:
```
┏━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━┓
┃ Parameter ┃ Best value ┃ Error ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━┩
│ g         │ -0.001     │ 0.331 │
└───────────┴────────────┴───────┘
[card.txt](https://github.com/user-attachments/files/17981415/card.txt)

```
[card.txt](https://github.com/user-attachments/files/17981426/card.txt)

**Comment by LucaMantani** [comment]
@jacoterh when you have the time, please do an independent check yourself, with a different card just for the sake of it, so that we feel confident enough to merge.

**Comment by giacomomagni** [comment]
> I tested for a fake UV model with non-linear constraints (card used in attachment) and it works fine: main:
> 
> [card.txt](https://github.com/user-attachments/files/17981426/card.txt)

okay thanks!

**Review by giacomomagni — APPROVED** [review]

**Comment by jacoterh** [comment]
I also found agreement between main and #96kXi

[card.txt](https://github.com/user-attachments/files/18023567/PR99_UV_main.txt)

**Review by jacoterh — APPROVED** [review]

**Comment by LucaMantani** [comment]
@jacoterh what about the basis_rotation?

**Comment by jacoterh** [comment]
Not sure why we are worried about the basis rotation actually, this branch gives the same result as main for the EWPO: coefficient_histo
